### PR TITLE
Enable parallel arma parameters search

### LINKED ISSUE #8231: Parallel search for auto-regressive models?
#### Is your feature request related to a problem? Please describe
I've recently come across a situation when I needed to do a lot of AR model searches for a considerable number of samples. There are built-in tools for that in statsmodels, such as `statsmodels.tsa.ar_model.ar_select_order`, but the problem is that they are not parallel.

I kind of have a prototype implementation for parallel model search which I could put a bit more effort into and submit a PR. Would this be a welcome addition for the repo or it is basically a design decision to not have a parallel model search implementation at that level?

#### Describe the solution you'd like
I can implement a parallel model search using `joblib`. doing a search in parallel would be optional and could be implemented as an additional flag for the aforementioned routine. 

#### Describe alternatives you have considered
Parallelization on the data level - doing multiple searches of orders for multiple data slices. 

#### Additional context
This is not a feature request so to say, more like a question about parallel model search, which we can just close if this is of no interest to the project.

### PR DESCRIPTION
**Pull request by sofeikov**
- [x] closes #8231
- [x] tests added, but may fail on github, see below.
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

This PR enables ARMA parameter search to be run in a parallelised way through `joblib`. This is achieved by introducing an additional flag `parallel` that is set to `False` by default. This approach proves to be faster for certain parameter ranges/dataset sizes. I can include benchmarks if needed.

Please, let me know if this is useful. If it is, I will 
1. amend the change log and it should be good to go. 
2. we need to decide what to do with `joblib` dep for github actions. We can just install it as an additional dependency for that particular action stage.

### CODE REVIEW

File: statsmodels/tsa/stattools.py [hunk 1]
```diff
@@ -1876,15 +1888,23 @@ def arma_order_select_ic(
...
-        for ma in ma_range:
-            mod = _safe_arma_fit(y_arr, (ar, 0, ma), model_kw, trend, fit_kw)
-            if mod is None:
-                results[:, ar, ma] = np.nan
-                continue
-
-            for i, criteria in enumerate(ic):
-                results[i, ar, ma] = getattr(mod, criteria)
+
+    def get_ic_for_order(ar: int, ma: int, criteria: str):
```
Comment (josef-pkt) [inline statsmodels/tsa/stattools.py:1892]: should be `criterion`, singular, otherwise it's confusing

File: statsmodels/tsa/stattools.py [hunk 2]
```diff
@@ -1876,15 +1888,23 @@ def arma_order_select_ic(
...
+        mod = _safe_arma_fit(y_arr, (ar, 0, ma), model_kw, trend, fit_kw)
+        if mod is None:
+            return np.nan
+        return getattr(mod, criteria)
+    
+    def fit_order(ar, ma):
+        return [get_ic_for_order(ar, ma, criteria) for criteria in ic]
```
Comment (josef-pkt) [inline statsmodels/tsa/stattools.py:1899]: this is not a good loop.
`get_ic_for_order`  fits the model.
So you are fitting the model for each ic criterion separately, repeating the same work each time.
This should get all criteria from one `fit`.

**Comment by sofeikov** [comment]
What is the best way to handle additional (optional) joblib dependancy here? Should I just included an additional `pip install` into the github actions template or just add it to `requirements.txt`? Thanks!

**Comment by bashtage** [comment]
At the top of the test file use

```python
try:
    import joblib
    HAS_JOBLIB = True
except ImportError:
    HAS_JOBLIB = False
```

Then  you can test missing joblib using

```
@pytest.mark.skipif(not HAS_JOBLIB):
def test_parallel_arma_search_joblib_unavailable(): # or something
# Check exception raised
```

You can also modify the parameterization to be False only if joblib is missing, of [True, Fales] when it is installed.

**Comment by sofeikov** [comment]
So I ended up using https://docs.python.org/3/library/pkgutil.html#pkgutil.find_loader to determine whether the package is installed and can be loaded. If I went for `ImportError`, then linting checks would not pass, because in that case `joblib` would be considered as an unused import.

**Pull request closed without merging**